src/hvm.rs

```diff
@@ -3790,198 +3790,221 @@ fn show_runtime_error(err: RuntimeError) -> String {
             stack.push(ask_arg(rt, term, 2));
           }
         }
         OP2 => {
-          stack.push(ask_arg(rt, term, 1));
-          stack.push(ask_arg(rt, term, 0));
+          let arg0 = ask_arg(rt, term, 0);
+          let arg1 = ask_arg(rt, term, 1);
+          stack.push(arg1);
+          stack.push(arg0);
         }
+        NUM => {}
         CTR | FUN => {
           let arity = rt.get_arity(get_ext(term));
           for i in (0..arity).rev() {
-            stack.push(ask_arg(rt, term, i));
+            let arg = ask_arg(rt, term, i);
+            stack.push(arg);
           }
         }
         _ => {}
       }
     }
+ }
 
-    let cont = expr(rt, term, &names);
-    if lets.is_empty() {
-      cont
-    } else {
-      let mut output = Term::Var { name: 0 };
-      for (i, (_key, pos)) in lets.iter().enumerate() {
-        // todo: reverse
-        let what = String::from("?h");
-        let name = names.get(&pos).unwrap_or(&what);
-        let nam0 = if ask_lnk(rt, pos + 0) == Era() { String::from("*") } else { format!("a{}", name) };
-        let nam1 = if ask_lnk(rt, pos + 1) == Era() { String::from("*") } else { format!("b{}", name) };
-        let expr = expr(rt, ask_lnk(rt, pos + 2), &names);
-        if i == 0 {
-          output = Term::Dup { nam0: name_to_u128(&nam0), nam1: name_to_u128(&nam1), expr: Box::new(expr), body: Box::new(cont.clone()) };
-        } else {
-          output = Term::Dup { nam0: name_to_u128(&nam0), nam1: name_to_u128(&nam1), expr: Box::new(expr), body: Box::new(output) };
-        }
-      }
-      output
+  struct DupStore {
+    stacks: HashMap<Ptr, Vec<bool>>,
+  }
+
+  impl DupStore {
+    fn new() -> DupStore {
+      DupStore { stacks: HashMap::new() }
+    }
+    fn get(&self, col: Ptr) -> Option<&Vec<bool>> {
+      self.stacks.get(&col)
+    }
+    fn pop(&mut self, col: Ptr) -> bool {
+      let stack = self.stacks.entry(col).or_insert_with(Vec::new);
+      stack.pop().unwrap_or(false)
+    }
+    fn push(&mut self, col: Ptr, val: bool) {
+      let stack = self.stacks.entry(col).or_insert_with(Vec::new);
+      stack.push(val);
     }
   }
 
-  fn expr(rt: &Runtime, term: Ptr, names: &HashMap<u128, String>) -> Term {
-    let mut stack = vec![StackItem::Term(term)];
+  fn readback(rt: &Runtime, term: Ptr, names: &mut HashMap<Ptr, String>, seen: &mut HashSet<Ptr>, dup_store: &mut DupStore) -> Term {
+    enum StackItem {
+      Term(Ptr),
+      Resolver(Ptr),
+      SUPResolverSome(Ptr, bool), // auxiliar case when in SUP does not have a DUP to evaluate
+      SUPResolverNone(Ptr), // auxiliar case when in SUP does not have a DUP to evaluate
+    }
+
     let mut output = Vec::new();
+    let mut stack = vec![StackItem::Term(term)];
+
     while !stack.is_empty() {
       let item = stack.pop().unwrap();
       match item {
-        StackItem::Resolver(term) => {
+        StackItem::Term(term) => {
+          debug_assert!(term != 0);
```

Comment:
It would be nice to add this check to some other functions while we don't refactor to a newtype.